reproschema/models/tests/test_activity.py

```diff
@@ -0,0 +1,125 @@
+from ..item import Item
+
+my_path = os.path.dirname(os.path.abspath(__file__))
+
+# Left here in case Remi and python path or import can't be friends once again.
+# sys.path.insert(0, my_path + "/../")
+
+# TODO
+# refactor across the different test modules
+activity_dir = os.path.join(my_path, "activities")
+if not os.path.exists(activity_dir):
+    os.makedirs(os.path.join(activity_dir))
+
+"""
+Only for the few cases when we want to check against some of the files in
+reproschema/tests/data
+"""
+reproschema_test_data = os.path.join(my_path, "..", "..", "tests", "data")
+
+
+def test_default():
+
+    """
+    FYI: The default activity does not conform to the schema
+    so  `reproschema validate` will complain if you run it in this
+    """
+
+    activity = Activity()
+    activity.set_defaults()
+
+    activity.write(activity_dir)
+    activity_content, expected = load_jsons(activity)
+    assert activity_content == expected
+
+    clean_up(activity)
+
+
+def test_activity():
+
+    activity = Activity()
+    activity.set_defaults("activity1")
+    activity.set_description("Activity example 1")
+    activity.set_pref_label("Example 1")
+    activity.set_preamble(
+        "Over the last 2 weeks, how often have you been bothered by any of the following problems?"
+    )
+    activity.set_citation("https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1495268/")
+    activity.set_image(
+        {"@type": "AudioObject", "contentUrl": "http://example.com/sample-image.png"}
+    )
+    activity.set_compute("activity1_total_score", "item1 + item2")
+
+    item_1 = Item()
+    item_1.set_defaults("item1")
+    # TODO
+    # probably want to have items/item_name be a default
+    item_1.set_URI(os.path.join("items", item_1.get_filename()))
+    # TODO
+    # We probably want a method to change those values rather that modifying
+    # the instance directly
+    item_1.skippable = False
+    item_1.required = True
+    """
+    Items are appended and this updates the  the ``ui`` ``order`` and ``addProperties``
+    """
+    activity.append_item(item_1)
+
+    item_2 = Item()
+    item_2.set_defaults("item2")
+    item_2.set_filename("item_two")
+
+    """
+    In this case the URI is relative to where the activity file will be saved
+    """
+    item_2.set_URI(os.path.join("..", "other_dir", item_2.get_filename()))
+    item_2.required = True
+    activity.append_item(item_2)
+
+    item_3 = Item()
+    item_3.set_defaults("activity1_total_score")
+    """
+    By default all files are save with a json.ld extension but this can be changed
+    """
+    file_ext = ""
+    item_3.set_filename("activity1_total_score", file_ext)
+    item_3.set_URI(os.path.join("items", item_3.get_filename()))
+    item_3.skippable = False
+    item_3.required = True
+    item_3.visible = False
+    activity.append_item(item_3)
+
+    activity.write(activity_dir)
+    activity_content, expected = load_jsons(activity)
+    assert activity_content == expected
```

Comment:
@Remi-Gau with respect to the below comment, instead of `assert` shouldn't this be validating the shape of the file using the SHACL validator?
if so, how to write `reproschema -l DEBUG validate <file-name.jsonld>` in this test file?

Reply:
You are correct that we should validate the shape. 

But we also want to make sure that we don't dump extra information in the jsonld that would not trigger an error in the validation: if user want to add extra stuff that's ok, but it should not beause of a side effect of our code.

I dug up how to call the validation directly: the things below seems to get the job done.

```python
from reproschema.validate import validate

fp = os.path.join(activity_dir, "activity2_schema.jsonld")
validate(None, fp)
```

Reply:
thanks for the above code. 
also, what did you mean by extra information? if the validation is in place, I don't think it would allow for any extra information (as the shape is a closed one).

Reply:
Will give an example: I want to add RRIDs to one of my list of preset response options. So I would just add them in the jsonld. (See below)

Currently the reproschema validation does not complain about this: if it should then we must change that.

If this is the expected behavior then, we just want to make sure that the code for protocol, activity, field creation does not add extra stuff and only sticks to things degined in the schema.

```
{
    "@context": "https://raw.githubusercontent.com/ReproNim/reproschema/1.0.0-rc4/contexts/generic",
    "@type": "reproschema:ResponseOption",
    "@id": "mri_softwares.jsonld",
    "valueType": "xsd:integer",
    "minValue": 0,
    "maxValue": 4,
    "choices": [
        {
            "name": {
                "en": "SPM"
            },
            "value": 0, 
            "RRID": "somethingg"       <------- Extra stuff not covered by schema
        },
        {
            "name": {
                "en": "FSL"
            },
            "value": 1,
            "RRID": 251        <------- Extra stuff not covered by schema
        }
    ]
}
```

Reply:
> shouldn't this be validating the shape of the file using the SHACL validator?

Actually I was tired yesterday when I replied. We don't have to check the shape of the stuff we generate:
as long as the jsonld that we test against have a valid shape, then checking that the content generated matches the expected one will be enough.

And as we do have to check the content to make sure our code behaves as expected, we don't need to rely on the shape validation.

Makes sense?

Reply:
how will you already have the JSON-LD content beforehand? I don't understand that.

Reply:
Well you have to sort of do that manually once to create the data you will run the output of your test against.

I think in this case I tried to reuse some of the test data that already exist in the [reproschema-py repo](https://github.com/ReproNim/reproschema-py/tree/master/reproschema/tests/data).

Reply:
Would it help if we had a quick meeting on the testing philosophy and how to run those tests?

Reply:
> Would it help if we had a quick meeting on the testing philosophy and how to run those tests?

Hey @Remi-Gau sorry I missed your message here. sure, let's meet tomorrow at whatever time works for you.